### PR TITLE
Faster path drawing for the cairo backend (cairocffi only)

### PR DESCRIPTION
**Pull request by anntzer**
When using cairocffi only:

Accelerate Path drawing by directly constructing a cairo_path_t using numpy operations rather than repeatedly calling the cairo API to add one point at a time.
Accelerate PathCollection drawing by concatenating Paths that have the same drawing parameters (color, etc., i.e. GraphicsContext attributes).
Minor additional performance improvements (remove an unneeded ndarray copy and an unneeded context save/restore pair).

Overall, this brings the performance of examples/mplot3d/wire3d_animation.py (used as a microbenchmark) from ~8.3fps to ~11.6fps on the same computer (just optimizing Paths instead of PathCollections gives ~10.5fps), using the gtk3cairo backend.  As a comparison point, the gtk3agg backend renders ~16.2fps for that example.

From profiling, I believe that further improvements (on that same microbenchmark) could be achieved by improving the text drawing (currently, we use the "toy" API, which is unfortunately the only one exposed by cairocffi so far; additionally, I am not certain whether ft2font exposes pointers to the underlying freetype structures).

Or (for other performance tests) draw_markers could probably also benefit from a similar approach as this PR.

### CODE REVIEW

File: lib/matplotlib/backends/backend_cairo.py [hunk 1]
```diff
@@ -76,6 +77,95 @@ def buffer_info(self):
...
+
+    # Fill the buffer.
+    buf = np.empty(cairo_num_data * 16, np.uint8)
+    as_int = np.frombuffer(buf.data, np.int32)
+    as_float = np.frombuffer(buf.data, np.float64)
+    mask = np.ones_like(as_float, bool)
+    as_int[::4][cairo_type_positions] = codes
+    as_int[1::4][cairo_type_positions] = cairo_type_sizes
+    mask[::2][cairo_type_positions] = mask[1::2][cairo_type_positions] = False
+    as_float[mask] = vertices.ravel()
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_cairo.py]: Could be re-arranged slightly to keep the `int` and `float` bits together.

File: lib/matplotlib/backends/backend_cairo.py [hunk 2]
```diff
@@ -76,6 +77,95 @@ def buffer_info(self):
...
+                ctx.move_to(*points)
+            elif code == Path.CLOSEPOLY:
+                ctx.close_path()
+            elif code == Path.LINETO:
+                ctx.line_to(*points)
+            elif code == Path.CURVE3:
+                cur = ctx.get_current_point()
+                ctx.curve_to(
+                    *np.concatenate([cur / 3 + points[:2] * 2 / 3,
+                                     points[:2] * 2 / 3 + points[-2:] / 3]))
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: previous version was actually incorrect (see basically https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Degree_elevation)..

File: lib/matplotlib/backends/backend_cairo.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: Not sure what you meant?

File: lib/matplotlib/backends/backend_cairo.py [hunk 3]
```diff
@@ -135,38 +225,16 @@ def _fill_and_stroke(self, ctx, fill_c, alpha, alpha_overrides):
...
-            clip = None
-
+        # Clip the path to the actual rendering extents if it isn't filled.
+        clip = (ctx.clip_extents()
+                if rgbFace is None and gc.get_hatch() is None
+                else None)
         transform = (transform
-                     + Affine2D().scale(1.0, -1.0).translate(0, self.height))
-
+                     + Affine2D().scale(1, -1).translate(0, self.height))
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: transforms are mutable but don't implement iadd (so += actually doesn't modify the transform in place), but I was too lazy to check and want to make clear that we're indeed not mutating the transform.

File: lib/matplotlib/backends/backend_cairo.py [hunk 4]
```diff
@@ -190,7 +258,7 @@ def draw_markers(
             filled = True
 
         transform = (transform
-                     + Affine2D().scale(1.0, -1.0).translate(0, self.height))
+                     + Affine2D().scale(1, -1).translate(0, self.height))
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: as above

File: lib/matplotlib/backends/backend_cairo.py [hunk 5]
```diff
@@ -218,6 +286,55 @@ def draw_markers(
...
+        def _draw_paths():
+            if not grouped_draw:
+                return
+            gc_vars, rgb_fc = reuse_key
+            gc = copy.copy(gc0)
+            # We actually need to call the setters to reset the internal state.
+            vars(gc).update(gc_vars)
+            for k, v in gc_vars.items():
+                try:
+                    getattr(gc, "set" + k)(v)
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: fixed

File: lib/matplotlib/backends/backend_cairo.py [hunk 6]
```diff
@@ -218,6 +286,55 @@ def draw_markers(
...
+            self._fill_and_stroke(
+                gc.ctx, rgb_fc, gc.get_alpha(), gc.get_forced_alpha())
+
+        for xo, yo, path_id, gc0, rgb_fc in self._iter_collection(
+                gc, master_transform, all_transforms, path_ids, offsets,
+                offsetTrans, facecolors, edgecolors, linewidths, linestyles,
+                antialiaseds, urls, offset_position):
+            path, transform = path_id
+            transform = (Affine2D(transform.get_matrix()).translate(xo, yo)
+                         + Affine2D().scale(1, -1).translate(0, self.height))
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: I don't.

File: lib/matplotlib/backends/backend_cairo.py [hunk 7]
```diff
@@ -218,6 +286,55 @@ def draw_markers(
...
+            # rgb_fc could be a ndarray, for which equality is elementwise.
+            new_key = vars(gc0), tuple(rgb_fc) if rgb_fc is not None else None
+            if new_key == reuse_key:
+                grouped_draw.append((path, transform))
+            else:
+                _draw_paths()
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_cairo.py]: The draw order is maintained, it's just that I batch the draws that use the same gc together.

**Comment by anntzer** [comment]
Comments addressed

File: lib/matplotlib/backends/backend_cairo.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/backends/backend_cairo.py]: Well, you start with a buffer, make an `int` version, then a `float` version, then a mask (which is for `float`). But then go back to filling the `int` version, modifying the mask (which is `float` stuff again), then fill the `float` version. So I mean more like:
```python
    buf = np.empty(cairo_num_data * 16, np.uint8)
    as_int = np.frombuffer(buf.data, np.int32)
    as_int[::4][cairo_type_positions] = codes
    as_int[1::4][cairo_type_positions] = cairo_type_sizes
    as_float = np.frombuffer(buf.data, np.float64)
    mask = np.ones_like(as_float, bool)
    mask[::2][cairo_type_positions] = mask[1::2][cairo_type_positions] = False
    as_float[mask] = vertices.ravel() 
```

File: lib/matplotlib/backends/backend_cairo.py [hunk 8]
```diff
@@ -76,6 +77,95 @@ def buffer_info(self):
...
+_CAIRO_PATH_TYPE_SIZES[cairo.PATH_CURVE_TO] = 4
+_CAIRO_PATH_TYPE_SIZES[cairo.PATH_CLOSE_PATH] = 1
+
+
+def _convert_path(ctx, path, transform, clip=None):
+    return _convert_paths(ctx, [path], [transform], clip)
+
+
+def _convert_paths(ctx, paths, transforms, clip=None):
+    return (_convert_paths_fast if HAS_CAIRO_CFFI else _convert_paths_slow)(
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/matplotlib/backends/backend_cairo.py]: By top-level, I meant `if HAS_CAIRO_CFFI: _convert_paths = _convert_paths_fast ...`

**Comment by anntzer** [comment]
comments addressed

**Comment by QuLogic** [comment]
Generally, I don't see a problem with this change, but using your microbenchmark, I can't seem to reproduce the speedup you see:

```
$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post1307.dev0+g84008ea05
Backend: TkAgg
Average FPS: 19.965329

$ python3 examples/mplot3d/wire3d_animation.py
/usr/lib/python3.6/site-packages/cairocffi/surfaces.py:651: UserWarning: implicit cast from 'char *' to a different pointer type: will be forbidden in the future (check that the types are as you expect; use an explicit ffi.cast() if they are correct)
  ffi.cast('char*', address), format, width, height, stride)
Matplotlib version: 2.1.1.post1307.dev0+g84008ea05
Backend: GTK3Agg
Average FPS: 20.217351

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post1307.dev0+g84008ea05
Backend: agg
Average FPS: 22.100502

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post1307.dev0+g84008ea05
Backend: GTK3Cairo
Average FPS: 12.131168

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post1307.dev0+g84008ea05
Backend: cairo
Average FPS: 70.399400
```
vs.
```
$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post18.dev0+gebcfd908c
Backend: TkAgg
Average FPS: 18.767531

$ python3 examples/mplot3d/wire3d_animation.py
/usr/lib/python3.6/site-packages/cairocffi/surfaces.py:651: UserWarning: implicit cast from 'char *' to a different pointer type: will be forbidden in the future (check that the types are as you expect; use an explicit ffi.cast() if they are correct)
  ffi.cast('char*', address), format, width, height, stride)
Matplotlib version: 2.1.1.post18.dev0+gebcfd908c
Backend: GTK3Agg
Average FPS: 19.497821

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post18.dev0+gebcfd908c
Backend: agg
Average FPS: 21.796568

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post18.dev0+gebcfd908c
Backend: GTK3Cairo
Average FPS: 12.018317

$ python3 examples/mplot3d/wire3d_animation.py
Matplotlib version: 2.1.1.post18.dev0+gebcfd908c
Backend: cairo
Average FPS: 68.990914
```

**Comment by anntzer** [comment]
Are you certain your gtk3cairo is using cairocffi rather than pycairo (which is not accelerated by this PR)?
I just tried and can still get the difference, ~8.7fps vs 12.2fps.

**Comment by QuLogic** [comment]
OK, I'm not really sure what happened, but I just tried again and I get 12.705427 vs. 17.470702 now.

**Comment by lazka** [comment]
Would it help you if pycairo had something like `cairo.Path.create_for_data(buffer, num_data)` ?

**Comment by anntzer** [comment]
I moved my efforts to just doing everything in C(++) (https://github.com/anntzer/mplcairo) but I guess such a method can't hurt in general...
Perhaps also worth (for you) to discuss this with cairocffi to present the same API in both projects.

**Comment by lazka** [comment]
> I moved my efforts to just doing everything in C(++) (https://github.com/anntzer/mplcairo) but I guess such a method can't hurt in general...

ok

> Perhaps also worth (for you) to discuss this with cairocffi to present the same API in both projects.

yeah, but it's unmaintained atm

**Pull request merged**